pios_dmashot: Initial implementation.

glowtape opened the pull request:
This is the DMA implementation of DShot. It is implemented in a self-contained module that hooks into pios_servo via #ifdefs on F4 targets.

You set up a configuration struct in board_hw_defs.c that specifies which timers can use what DMA channel and stream combination (among other things), which then gets fed to PIOS_DMAShot_Init(). Timers that don't have a DMA configuration will fall back to GPIO DShot within pios_servo. Optionally you can slave DMA writes to a single CC register onto another timer, to deal with things like Seppuku's TIM14.

Each timer can run on its own DShot frequency, so you can mix DShot300 - DShot1200, if you want to, for whatever reason.

As per @mlyle's suggestion, it doesn't use the prescaler and calculates the period and duty cycles during initialization. The duty cycles chosen are the out-of-spec ones BF/BLHeli uses. They work with KISS though.

As far as memory use goes, if DMAShot is compiled in (data, not code), it uses statically:
- 4 bytes to store a pointer to the DMA configuration.
- 4 bytes to allocate an array to.

If DMAShot gets initialized with a configuration, it'll eventually allocate 32 bytes to aforementioned array, should an user configure DShot on any timer, to hold up to 8 pointers to the internal timer configuration allocations.

For each timer configured, it'll allocate 42 bytes on a struct to store a variety of data to look up.

For DMA buffers, it'll allocate at word (32-bit) or half-word (16-bit) where appropriate. 18 "bits" will be allocated, a leading and trailing one that silences the timer, and 16 actual "bits". So the buffer would be at least 36 or 72 bytes (multiplied by amount of channels per timer).

Due to the nature of DMAR writes, there might be some overallocation. If there's only two pins on a timer, but they're spaced apart as much as possible (one pin on channel 1 and another on channel 4), it has to allocate a DMA buffer for four channels (up to 288 bytes on a 32-bit timer) to fill in for in-between. If two pins are only separated by one or when pins are adjacent, memory is allocated accordingly, i.e. always the minimum.

Best situation would be four motors on a single 16-bit timer, that'd use 226 bytes of RAM (excluding code). The RE1 would use 370 bytes, because it has the first four pins on a single but 32-bit timer. The Seppuku would use 418 bytes, due four pins being spread across three but 16-bit timers. (If I got my math correctly.)

I could successfully drive DShot on all eight pins of the RE1 and Seppuku, at mixed frequencies. OG Brain was only tested on the first four, because I don't have any 8-pin JST cables anymore.

Still need to add other targets.

Old PR message:

> PR created for general appraisal and critique. Given how much I chased my tail over a bunch of stupid bugs, I'm going to leave it a day or two alone.
> 
> Everything is pretty much standalone with a bunch of hooks into pios_servo.
> 
> Tested on a BrainRE1, I can drive all eight pins with DMA DShot (specifically disabled bitbanging code to check).
> 
> Things I need to do:
> - Implement actual frequency selection. Maybe compute it per timer instead of a bunch of #ifdef'd defines (like currently for F446).
> - Test whether 16-bit timers can do away with half-word writes, to save memory in the DMA buffers. Full word writes to 16-bit timers (e.g. TIM1 and TIM8 on RE1) don't break anything, though.
> - Add code to deal with TIM14 for Seppuku (write to DMAR, if it has it, or direct write to CCR, slaved to other timer like second TIM3_UP).

glowtape committed:
pios_dmashot: Initial implementation.

mlyle commented:
> Add code to deal with TIM14 for Seppuku (write to DMAR, if it has it, or direct write to CCR, slaved to other timer like second TIM3_UP).

I think you'll need to write to the CCR directly, btw.  I don't know the exact logic for the generation of DMA requests beats and for address incrementing / DMAR magic in the hardware, but if they're not completely independent it's going to cause a problem using another timer's DMA request line.  (And even if they are independent on F405, it seems safer to be explicit).

tracernz commented on flight/PiOS/STM32F4xx/inc/pios_dmashot.h:
```diff
@@ -0,0 +1,78 @@
+#define DSHOT600_FREQUENCY						(DSHOT300_FREQUENCY*2)
+#define DSHOT900_FREQUENCY						(DSHOT300_FREQUENCY*3)
+#define DSHOT1200_FREQUENCY						(DSHOT300_FREQUENCY*4)
+#define DSHOT150_FREQUENCY						(DSHOT300_FREQUENCY/2)
```
Missing spaces around operators

tracernz commented on flight/PiOS/STM32F4xx/inc/pios_dmashot.h:
```diff
@@ -0,0 +1,78 @@
+
+void PIOS_DMAShot_FireDMAs();
+
+bool PIOS_DMAShot_GreenLight();
```
This is not a good function name IMO. We should stick to formal language.

tracernz commented on flight/PiOS/STM32F4xx/inc/pios_dmashot.h:
```diff
@@ -0,0 +1,78 @@
+
+bool PIOS_DMAShot_WriteValue(const struct pios_tim_channel *servo_channel, uint16_t throttle);
+
+void PIOS_DMAShot_FireDMAs();
```
Marginal name.

tracernz commented on flight/PiOS/STM32F4xx/inc/pios_dmashot.h:
```diff
@@ -0,0 +1,78 @@
+
+bool PIOS_DMAShot_GreenLight();
+
+#endif // PIOS_DMASHOT_H
```
Text files require newlines to terminate all character sequences.

tracernz commented on flight/PiOS/STM32F4xx/inc/pios_dmashot.h:
```diff
@@ -0,0 +1,78 @@
+
+};
+
+void PIOS_DMAShot_Init(const struct pios_dmashot_cfg *config);
```
Please add Doxygen documentation for all public interfaces.

mlyle replied:
Ditto

mlyle replied:
(Oh, I see they are in the C file.  Can you move them to the header, since this is the preferred way now?)

mlyle replied:
This is still a blocker before merge.

tracernz commented on flight/PiOS/STM32F4xx/pios_dmashot.c:
```diff
@@ -0,0 +1,423 @@
+#define TIMC_TO_INDEX(c)			((c)>>2)
+#define NUM_CHANNELS(s)				(TIMC_TO_INDEX((s)->high_channel - (s)->low_channel) + 1)
+
+#define DMA_DISABLE(s)				{ DMA_Cmd(s, DISABLE); while (DMA_GetCmdStatus(s) == ENABLE); }
```
Hiding tight loops in macros harms readability/analysis IMO.

tracernz commented on flight/PiOS/STM32F4xx/pios_dmashot.c:
```diff
@@ -0,0 +1,423 @@
+		// Allocate memory
+		servo_timers = PIOS_malloc_no_dma(sizeof(struct servo_timer) * MAX_TIMERS);
+		if(!servo_timers)
+			ERROR();
```
This hurts readability IMO. Does it return, does it assert? Makes life harder when chasing bugs.

tracernz commented on flight/PiOS/STM32F4xx/inc/pios_dmashot.h:
```diff
@@ -0,0 +1,78 @@
+
+#ifdef STM32F446xx
+
+#define DSHOT150_FREQUENCY						11250000
```
Most of these belong in a private header, or the implementation (.c), not a public header (public headers should be restricted to minimal public interface only).

tracernz commented on flight/PiOS/STM32F4xx/pios_dmashot.c:
```diff
@@ -0,0 +1,423 @@
+bool PIOS_DMAShot_Enabled()
+{
+	return dmashot_cfg != NULL;
+}
```
Missing newline.

tracernz reviewed (commented):
There a lot of code style issues in pios_dmashot.*. Please run uncrustify when you've finished writing it. :wink:

glowtape committed:
seppuku: Add DMA DShot support.

mlyle commented:
```
/var/lib/jenkins/jobs/dron/workspace@2/Nodes/master/flight/PiOS/STM32/pios_servo.c:46:26: fatal error: pios_dmashot.h: No such file or directory
compilation terminated.
```

glowtape commented:
Yeah I know, the DMA DShot stuff currently in PIOS/STM32F4xx. I have to move it outta there. It should compile with the F4 headers regardless of what the actual target is?

glowtape committed:
brainre1: Add DMAShot support.

glowtape committed:
brain: Add DMAShot support.

glowtape committed:
pios_dmashot: How about we don't overallocate the pointer array.

glowtape commented:
Updated original message.

glowtape committed:
pios_dmashot: When fixing the overallocation, fixing the zeroing might be great.

glowtape committed:
pios_dmashot: Remove timer pointer, because redundant.

The DMA config struct has stored a timer ID. A timer only gets registered when it matches a DMA configuration, and since that one gets referred, we might as well just use that one. Sysclock = 0 to disable it regardless.

glowtape committed:
pios_dmashot: Only run OCInit on used channels.

glowtape committed:
pios_dmashot: Faulty logic disables servo on timing changes.

glowtape committed:
pios_dmashot: Use STM32 #defines to configure master timer.

glowtape committed:
pios_dmashot: Start master timer after signal timer.

glowtape committed:
seppuku: Drop DMAShot #defines for the STM32 ones.

glowtape committed:
pios_dmashot: Don't keep calling a function unnecessarily.

glowtape committed:
pios_servo: Bound the position before implicitly casting it to uint16_t.

Might cause funny results on negative values, that probably shouldn't happen to begin with.

glowtape committed:
pios_dmashot: Only allocate DMA buffers when there's servos configured.

glowtape commented:
With TBS Crossfire receiver, Autotune (w/ decimation on RE1) enabled, memory free (heap/fastheap) for SyncPWM or DShot, total memory usage when enabled:

All pins to DShot:
- Seppuku: SyncPWM (29136/51436), DShot (28740/51276) = 556 bytes.
- BrainRE1: SyncPWM (16864/0), DShot (16236/0) = 628 bytes.
- OG Brain: SyncPWM (28236/51648), DShot (27804/51568) = 512 bytes.

At least pins 1-4 to DShot:
- Seppuku: SyncPWM (29136/51436), DShot (28884/51316) = 372 bytes.
- BrainRE1: SyncPWM (16864/0), DShot (16536/0) = 328 bytes.
- OG Brain: SyncPWM (28236/51648), DShot (27948/51608) = 328 bytes.

glowtape commented:
Flew two packs of this at DShot600, three packs at DShot300 (OSD was not wobbly, so no accidental GPIO version). Seemed to work fine. Was too cold today, so I quit eventually. DShot1200 works on the bench using the Test Outputs tab, but I'm still going to fly it soon.

mlyle commented:
@Glowtape Awesome!

What FC was the flight test on?

glowtape commented:
A brand-new Seppuku.

mlyle commented:
Is this still WIP or are you ready for review?

glowtape committed:
pios_dmashot: Fix some comments.

glowtape committed:
pios_dmashot: No point in a bool return value in _WriteValue().

glowtape commented:
Now it's ready for review.

mlyle commented on flight/PiOS/STM32/pios_servo.c:
```diff
@@ -375,9 +386,21 @@ int PIOS_Servo_SetMode(const uint16_t *out_rate, const int banks, const uint16_t
+			ret = 1;
+#if defined(PIOS_INCLUDE_DMASHOT)
+			if (PIOS_DMAShot_IsConfigured()) {
+				uint32_t freq = rate == SHOT_DSHOT1200 ? DMASHOT_1200 : rate == SHOT_DSHOT600 ? DMASHOT_600 : DMASHOT_300;
```
No ternary-to-death, please?  Any other structure is OK

(or if going to use ternary, please indent it like other uses of ternary-as-case)

mlyle commented on flight/PiOS/STM32/pios_servo.c:
```diff
@@ -391,7 +414,18 @@ int PIOS_Servo_SetMode(const uint16_t *out_rate, const int banks, const uint16_t
 				case SHOT_DSHOT600:
 				case SHOT_DSHOT1200:
+#if defined(PIOS_INCLUDE_DMASHOT)
+					if (PIOS_DMAShot_IsConfigured() && PIOS_DMAShot_RegisterServo(chan)) {
```
Important test case: PIOS_INCLUDE_DMASHOT on a target with no dma channels configured (To verify fallback logic)

glowtape replied:
I already had a partial confirmation, when I was testing on Seppuku without slaving TIM14, resulting in mixed DMA and GPIO DShot. Just tested by commenting out PIOS_DMAShot_Init() in pios_board.c on Seppuku while leaving the #define intact. Motors started spinning.

Pudding for mixed DShot sources (before I changed call order in pios_servo). Channel 2 is GPIO, rest is DMA:
![dualshot](https://cloud.githubusercontent.com/assets/4010813/25203925/80ca04f6-255b-11e7-81e4-651353756abf.png)

mlyle commented on flight/PiOS/STM32/pios_servo.c:
```diff
@@ -460,6 +503,13 @@ void PIOS_Servo_SetFraction(uint8_t servo, uint16_t fraction,
 			return;
+		case SYNC_DSHOT_DMA:
+#if defined(PIOS_INCLUDE_DMASHOT)
+			if (PIOS_DMAShot_IsConfigured()) {
```
This is for F0/flyingpio.  I wonder if it can be made to work with this. :D

glowtape replied:
If we wrap DMA_Cmd in a macro and place an #ifdef in _DMASetup(), it can probably work. For F0 and also F3 I guess. I don't have any of those for testing, though.

mlyle commented on flight/PiOS/STM32F4xx/pios_dmashot.c:
```diff
@@ -0,0 +1,591 @@
+
+// This is to do half-word writes where appropriate. TIM2 and TIM5 are 32-bit, the rest
+// are 16-bit. Can do full word writes to 16bit regardless, but it doubles DMA buffer size
+// unnecessarily.
```
And to confirm-- you can't do half-word writes to the 32 bit timer?

glowtape replied:
Just checked again, doesn't work. RE1 has pin 1-4 on TIM5, which is 32bit. When I try to do half-word DMA into the DMAR of a 32bit timer, it doesn't seem to execute the transfer.

mlyle replied:
OK, thx.

mlyle commented on flight/PiOS/STM32F4xx/pios_dmashot.c:
```diff
@@ -0,0 +1,591 @@
+const struct pios_dmashot_cfg *dmashot_cfg;
+struct servo_timer **servo_timers;
+
+static inline int PIOS_DMAShot_GetFIFOCadence(int num_chan)
```
What's the reasoning for this?

glowtape replied:
From the docs: "The DMA burst size on the memory port and the FIFO threshold configuration must match."

mlyle replied:
Yes, I'm more saying.. why are you varying the DMA burst size on the memory port?  It's not related to the dma request / transfer burst length.

glowtape replied:
Seems like I've misunderstood the docs then. Reviewing the table of burst configurations, I can rarely make it fit anyway (esp. since DMA buffer size depends on channels used per timer), which implies it shouldn't have worked, yet it didn't blow.

And I just tested it with FIFO disabled, there's still a proper DShot signal on the wire regardless (on both Seppuku and RE1, which have different F4). So this gets scrubbed I guess.

mlyle replied:
They need to be chosen from the table to be a legal transfer for the **memory burst size** and element size.  But the memory burst size doesn't need to match the peripheral operation size, and the burst thing that the timer does is different from a DMA controller burst anyways.

Basically, every time the FIFO isn't filled, the DMA tries to get control of the memory bus, and do a sequence of N transfers in order to fill the fifo up to the desired level.  Because turning around the memory bus takes time and sequential transfers are faster, this uses the memory bandwidth more efficiently.  Some combinations are not legal (e.g. ones that risk overfilling the FIFO)

I believe at least-- this is my interpretation.

mlyle commented on flight/PiOS/STM32F4xx/pios_dmashot.c:
```diff
@@ -0,0 +1,591 @@
+		if (!s_timer || !s_timer->sysclock)
+			continue;
+
+		if (s_timer->dma->timer == servo_channel->timer) return s_timer;
```
line break before return

mlyle commented on flight/PiOS/STM32F4xx/pios_dmashot.c:
```diff
@@ -0,0 +1,591 @@
+		for (int i = 0; i < MAX_TIMERS; i++) {
+			if (!servo_timers[i]) {
+				s_timer = PIOS_malloc_no_dma(sizeof(struct servo_timer));
+				memset(s_timer, 0, sizeof(struct servo_timer));
```
I'm ignoring allocation / memset for now.. in my next review pass I'll look at them closely as they require a focused review.

mlyle replied:
(though should s_timer be checked for NULL? )

mlyle commented on flight/PiOS/STM32F4xx/pios_dmashot.c:
```diff
@@ -0,0 +1,591 @@
+
+		int f = s_timer->sysclock / s_timer->dshot_freq;
+
+		s_timer->duty_cycle_0 = f * DSHOT_DUTY_CYCLE_0 / 100;
```
Suggest + 50 in these

glowtape replied:
Not sure I understand.

mlyle replied:
(f * DSHOT_DUTY_CYCLE_0 + 50) / 100

Basically, get correct rounding when you can't hit timing exactly.

mlyle commented on flight/PiOS/STM32F4xx/pios_dmashot.c:
```diff
@@ -0,0 +1,591 @@
+			continue;
+
+		DMA_Cmd(s_timer->dma->stream, DISABLE);
+		while (DMA_GetCmdStatus(s_timer->dma->stream) == ENABLE) ;
```
I suggest we wait for completion --before disabling--, so that if jitter causes us to show up when another update is in progress we wait for it to be done.

mlyle replied:
(Alternatively we could drop the update entirely and return)

glowtape replied:
Added a check for the transfer complete flag. If it's not been set, it'll drop the update. Probably better that way, than to fear the erratic wrath of the watchdog.

mlyle replied:
Message lengths are very short compared to watchdog intervals (e.g. that's why we get away with bit-banging them).

Arguments for blocking: might get the useful update out sooner. If DMA/something gets stuck, we actually **do** trigger the watchdog instead of just returning immediately.

Arguments against: Prevent a chain of 2-3 updates being slightly late; instead absorb the penalty all at once.

glowtape replied:
Here's the OG Brain, which is already jittery with a lean configuration, at DShot300, with autotune and 250hz fullbore logging enabled to OpenLager at 2470000 baud (picked the worst case found):

![jitter2](https://cloud.githubusercontent.com/assets/4010813/25207959/df86e520-2571-11e7-8297-206f2391d82d.png)

Looks like an argument for dropping (actually, this screenshot is already from the build with the check for transfer complete flag). Do we expect DMA to break?

mlyle replied:
Wow.. OG brain doing so badly at 1KHz gyro is ... not great, but not unexpected given I2C i guess. 

if updates come nominally at 1ms, intervals, and you get one at 0ms, 1.99ms (should be 1ms), 2.01ms, 3.00ms, say..

If you drop, from 0-1.99ms you use the value from 0ms, from 1.99ms-3ms you use the value from 1ms,  and at 3ms you use the value from 3ms.

If you busy-wait, from 0-1.99ms you use the value from 0ms, from 1.99 to say 2.05ms you use the value from 1ms, and from 2.05ms to 3ms you use the value from 2ms.

> Do we expect DMA to break?

It shouldn't-- It's just a question of what makes any potential defect less severe (if someone stomps a timer configuration or whatever).

glowtape replied:
OK, changed it to wait for the transfers to finish.

mlyle commented:
Sorry for not bundling comments into a review--- by the time I realized I had screwed up I was "committed" heh..

This is a very good start.  I need to check the memory stuff in depth in a separate pass.

glowtape committed:
pios_dmashot: Fixes.

- Skip DMA update, if transfer not complete.
- Less ugly DShot frequency selection.
- Style.

glowtape committed:
pios_dmashot: Disable FIFO (not needed), and remove faulty FIFO threshold selection.

glowtape committed:
pios_dmashot: Improve rounding for DShot duty cycle calc.

tracernz commented on flight/PiOS/STM32/pios_servo.c:
```diff
@@ -389,7 +389,19 @@ int PIOS_Servo_SetMode(const uint16_t *out_rate, const int banks, const uint16_t
-				uint32_t freq = rate == SHOT_DSHOT1200 ? DMASHOT_1200 : rate == SHOT_DSHOT600 ? DMASHOT_600 : DMASHOT_300;
+				uint32_t freq;
+				switch(rate) {
+					default:
```
default case should never happen? Should be an assert.

glowtape committed:
pios_dmashot: More fixes.

- Re-enabled FIFO, at full threshold.
- Allocate buffers 16-byte aligned, for DMA burst transfers.
- Assert in pios_servo, when an unknown DShot mode has been encountered.

glowtape committed:
pios_dmashot: Wait for DMA transfers to finish instead of dropping updates.

glowtape commented:
Left the OG Brain running at 90+ CPU for 45 minutes, waiting for DMA to break or delay things so much that the watchdog kicks in on the transfer complete waiting loop. Nothing happened. (Had to flash firmware to go flying, otherwise I'd have left it running.)

Looked at the output occasionally, the shortest distance I spotted was 78us between updates. Nothing that looked like an outright collision, though.

![collision](https://cloud.githubusercontent.com/assets/4010813/25237453/037517da-25eb-11e7-8047-f24d329d1c3f.png)

glowtape committed:
pios_dmashot: Fix bootloop when trying to change output when DMAShot is active.

glowtape commented:
Flew all my packs on the latest of this configured to DShot1200, on my Seppuku quad, together with #1714 and #1642 configured to extended prediction period. No issues.

mlyle commented on flight/PiOS/STM32F4xx/pios_dmashot.c:
```diff
@@ -0,0 +1,603 @@
+	if (!found) {
+		for (int i = 0; i < MAX_TIMERS; i++) {
+			if (!servo_timers[i]) {
+				s_timer = PIOS_malloc_no_dma(sizeof(struct servo_timer));
```
sizeof(*s_timer) is preferred for these

mlyle commented on flight/PiOS/STM32F4xx/pios_dmashot.c:
```diff
@@ -0,0 +1,603 @@
+
+		if (!s_timer->buffer.ptr) {
+			// Allocate buffer at timer resolution.
+			s_timer->buffer.ptr = PIOS_DMAShot_AllocateBuffer(
```
This factoring is much nicer.

glowtape committed:
pios_dmashot: Change what sizeof statements refer to.

glowtape committed:
pios_dmashot: Move doxygen stuff to headers.
